File: Biweekly Contest 126/100209. Mark Elements on Array by Performing Queries.go
package main
import (
	"container/heap"
	
)

type pair struct {
	value int
	index int
}

type minHeap []pair

func (h minHeap) Len() int            { return len(h) }
func (h minHeap) Less(i, j int) bool  { return h[i].value < h[j].value }
func (h minHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(x interface{}) { *h = append(*h, x.(pair)) }
func (h *minHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func unmarkedSumArray(nums []int, queries [][]int) []int64 {
	lenNums := len(nums)
	ind := make(map[int]bool)
	pq := make(minHeap, 0)
	heap.Init(&pq)
	var sumOfNums int64
	for i := 0; i < lenNums; i++ {
		ind[i] = true
		heap.Push(&pq, pair{value: nums[i], index: i})
		sumOfNums += int64(nums[i])
	}
	h := make(map[int][]int)
	for i := 0; i < lenNums; i++ {
		val := nums[i]
		if _, exists := h[val]; !exists {
			h[val] = []int{i}
		} else {
			h[val] = append(h[val], i)
		}
	}

	lent := len(queries)
	ans := make([]int64, lent)
	for i := 0; i < lent; i++ {
		left := queries[i][0]
		right := queries[i][1]

		

		if ind[left] {
			delete(ind, left)
			sumOfNums -= int64(nums[left])
		}

		for right > 0 {
			if len(pq) > 0 {
				minVal := heap.Pop(&pq).(pair)
				indexes := h[minVal.value]
				minInd := indexes[0]
				h[minVal.value] = indexes[1:]

				if ind[minInd] {
					right--
					delete(ind, minInd)
					sumOfNums -= int64(nums[minInd])
				}
			} else {
				break
			}
		}

		ans[i] = sumOfNums
	}

	return ans
}

File: Biweekly Contest 126/100209. Mark Elements on Array by Performing Queries.java

class Solution {

    public long[] unmarkedSumArray(int[] nums, int[][] queries) {
        int len = nums.length;
        HashSet<Integer> ind = new HashSet<>();
        PriorityQueue<Integer> pq =new PriorityQueue<>();
        long sumOfNums= 0 ;
        for(int i=0;i<len;i++){
            ind.add(i);
            pq.add(nums[i]);
            sumOfNums+=nums[i];
        }
        int lent = queries.length;
        HashMap<Integer,Queue<Integer>> h = new HashMap<>();
        for(int i=0;i<len;i++){
            int val = nums[i];
            if(h.get(val)==null){
                Queue<Integer> q = new LinkedList<>();
                q.add(i);
                h.put(val,q);
            }else{
                Queue<Integer>  q= h.get(val);
                q.add(i);
            }
        }
        long ans[] = new long[lent];
        for(int i=0;i<lent;i++){
            int left = queries[i][0];
            int right = queries[i][1];
            
            long sum = 0;

            

            if(ind.contains(left)){

                ind.remove(left);

                sumOfNums-=nums[left];

            }

            

            while(right>0){

                

                if(pq.size()>0){

                    

                    int minval = pq.poll();

                    Queue<Integer> indexes = h.get(minval);

                    int minInd = indexes.remove();

                    

                    if(ind.contains(minInd)){

                        right--;

                        ind.remove(minInd);

                        sumOfNums-=nums[minInd];

                    }else{

                        

                    }

                    

                }else

                {

                    break;

                }

                 

            }

            ans[i] = sumOfNums;

        }

        

        return ans;

        

    }

} 